packages/selection/src/useSelection.js

```diff
@@ -0,0 +1,180 @@
+      return { ...state, focusedIndex: 0 };
+    case ACTIONS.END:
+      return { ...state, focusedIndex: action.payload - 1 };
+    case ACTIONS.MOUSE_SELECT:
+      return {
+        ...state,
+        selectedIndex: action.payload,
+        focusedIndex: undefined
+      };
+    case ACTIONS.KEYBOARD_SELECT:
+      return { ...state, selectedIndex: action.payload };
+    case ACTIONS.EXIT_WIDGET:
+      return { ...state, focusedIndex: undefined };
+    default:
+      return state;
+  }
+}
+
+function useSelection({
+  defaultFocusedIndex,
+  direction = DIRECTION.HORIZONTAL,
+  selectedIndex,
+  focusedIndex,
+  onStateChange,
+  defaultRefKey = 'ref'
+} = {}) {
+  const [state, dispatch] = useReducer(stateReducer, {
+    focusedIndex: focusedIndex || defaultFocusedIndex,
+    selectedIndex: selectedIndex
+  });
+
+  const refs = [];
+  let numItems = 0;
+
+  useEffect(
+    () => {
+      if (state.focusedIndex !== undefined) {
+        refs[state.focusedIndex] && refs[state.focusedIndex].current.focus();
+      }
+    },
+    [state.focusedIndex]
+  );
+
+  const getContainerProps = ({ role = 'listbox', ...other } = {}) => ({
+    role,
+    ...other
+  });
+
+  const getItemProps = ({
+    key,
+    refKey = defaultRefKey,
+    selectedAriaKey = 'aria-selected',
+    role = 'option',
+    onFocus,
+    onKeyDown,
+    onClick,
+    ...other
+  } = {}) => {
+    if (typeof key === 'undefined') {
+      throw new Error(
+        '"key" must be defined within getItemProps regardless of being used within a .map()'
+      );
+    }
+
+    const currentIndex = numItems;
+    const isFocused =
+      state.focusedIndex === undefined
+        ? currentIndex === state.selectedIndex
+        : currentIndex === state.focusedIndex;
+    const isSelected = currentIndex === state.selectedIndex;
+
+    let tabIndex = -1;
+
+    if (!isFocused && !isSelected) {
+      if (currentIndex === defaultFocusedIndex) {
+        tabIndex = 0;
+      } else if (defaultFocusedIndex === undefined && currentIndex === 0) {
+        tabIndex = 0;
+      } else if (defaultFocusedIndex === -1) {
+        // Figure out when it's the last child and set tabIndex to 0 for that
+      }
+    } else {
+      tabIndex = isFocused ? 0 : -1;
+    }
+
+    refs[currentIndex] = createRef(null);
+
+    numItems++;
+
+    return {
+      role,
+      tabIndex,
+      [refKey]: refs[currentIndex],
+      [selectedAriaKey]: isSelected,
+      onFocus: composeEventHandlers(onFocus, () => {
+        dispatch({ type: ACTIONS.FOCUS, payload: currentIndex });
+      }),
+      onBlur: e => {
+        if (e.target.tabIndex === 0) {
+          dispatch({ type: ACTIONS.EXIT_WIDGET });
+        }
+      },
+      onClick: composeEventHandlers(onClick, () => {
+        dispatch({ type: ACTIONS.MOUSE_SELECT, payload: currentIndex });
+      }),
+      onKeyDown: composeEventHandlers(onKeyDown, e => {
+        if (
+          (e.keyCode === KEY_CODES.UP && direction === DIRECTION.VERTICAL) ||
```

Comment:
Could this logic be moved to the `stateReducer`? Have a single `KEYBOARD_NAV` event and use the keyCode as additional data to pass.

Not sure if that would be any cleaner though, especially with the need to `preventDefault()`. Not sure what's the best approach for this.